### ISSUE TITLE
Petitioner: Edit and display personal contact data

### ISSUE THREAD

**Issue by JessicaMarine** [flexion/ef-cms#150]
As a Petitioner, in order to ensure the court has my up-to-date contact information for document service, I need the ability to edit and display my personal contact data within my case.

## Pre-Conditions:
* Petitioner has successfully created an account and has personal data in the system (CMS)
* Petitioner has a username and password saved in the system (Cognito)

## Acceptance Criteria:
* Petitioner has the ability to modify their address and phone number
* Petitioner’s address, phone number are saved in the system (CMS)
* Petitioner’s address, phone number display in the system 
* Record of data change is captured through creation of a new docket entry that should read: "Notice of Change of Address by [name of person]" 
* Petitioner may not change the contact information for another party. THey should be instructed to contact the Court by filing a Form 10, Notice of Change of Address, available at https://ustaxcourt.gov/forms/NOCOA_Form_10.pdf

## Notes
* Changes to personal data do not change/affect cover sheet
* Practitoner and Sr. Attorney contact info changes in #1936
* For email change, going to explore if integration is possible and what just linking to Cognito looks like. (6/20)
* Do not display petitioner's personal data in the docket entry. Keep language simple, e.g. "It has changed." (6/20)
* When this happens, start a work item in the docket record's work queue. Other parties need to be informed via service. Generate a Change of Address and serve on parties. Paper and Electronic. Docket will link to a notice. Add this to the Notice epic.  NOT PART OF THIS STORY. WILL CREATE NEW STORY(6/20) #2146
* Petitioner could have multiple cases, each with a different address. (6/20)
* Moved changing email address to another story #2223 

## Tasks
- [x] Edit Button and Modals (Matt #2263)
- [x] Update user data on db (Matt #2269)
- [x] Validate contact info before saving (Joe #2289)
- [x] Show confirmation when saved (Joe #2285)
- [x] Authorize for write (isAuthorized->owner) (Eric)
- [x] Generate docket entry (Eric)
- [x] Write test scenarios (Mariah/Eric #2280)

## Definition of Done
**Product Owner**
 - [ ] Acceptance criteria have been met
**UX**
  - [x] Usability has been validated (Sarah)
**Engineering**
  - [x] Field level and page level validation errors (front-end and server side) integrated and functioning (Cody)
  - [x] Run the browser stack scripts to verify that application is usable in Chrome, Firefox, Safari, Edge and Internet Explorer (external only) (Matt)
  - [x] No errors reported in manual a11y testing (Eric)
  - [x] All new functionality verified to work with keyboard and osx voiceover (Eric)
  - [x] Code refactored for clarity and to remove any known technical debt (Eric)
  - [x]  README, other appropriate docs, and swagger/APIs fully updated - all code functions documented with jsdocs (Cody)
  - [x] UI should be touch optimized and responsive for external only(functions on supported mobile devices and optimized for screen sizes as required) (Eric)
  - [x] Library/module dependencies are up-to-date and are at the latest resolvable version (Eric)
  - [x] Code is deployed to DEV
  - [x] Errors in sonarcloud are fixed (https://sonarcloud.io/organizations/flexion-github/projects) (Joe)
  - [x] Alerting and monitoring for new functionality in place and successfully exercised (Eric)

__________
Outdated based on clarification (6/18):
Before Petitioner can file an electronic petition with the Court, they need to create an account.  Petitioner navigates to the webpage where they can create an account.  As part of the registration process, Petitioner will use an email address as a user name and select a secure password.  Petitioner will receive email confirmation of account creation. Petitioner validates email address.

When a Petitioner creates a new case within the CMS, they will enter additional contact information name, address, phone.  That information should be saved to the Petitioner's profile.  A Petitioner should be able to manage their profile at any time.

**Comment by ericsorenson** [comment]
Cognito's hosted UI doesn't allow password or email address modifications, so we'll need to use the Cognito API for these features.

**Comment by sarahridgway** [comment]
Note to update state drop down from full name to abbreviation on all US address field sets

**Comment by sarahridgway** [comment]
*Petitioner: Updating Contact Information*

* Petitioner may have multiple cases, each with their own unique name and contact information

*Contact Information Tab*

Petitioner change the contact information for each case individually through the Case Information tab on each Case Detail page.

* Clicking "Edit" under the Petitioner's contact information will open the Contact Info Modal
* Clicking "Why can't I edit this?" under secondary Petitioner's contact information will open Secondary Petitioners' Contact Info Modal  

![image.png](https://images.zenhubusercontent.com/5be0bad045b0290c78fb7458/9fe56445-9966-4867-a36e-2d962b56e130)

*Contact Info Modal*

* Display the contact info form for the case's primary petitioner petitionerContact type
* Pre-populate the contact form with the current information saved to the case
* User can edit any field
* Required fields must be filled in otherwise show standard errors for this form
* Allow the user to delete info from optional fields without throwing an error
* Global Change: state list to be updated to abbreviations ONLY for ALL U.S. address templates
* Clicking "Save" commits information to the case and closes the modal window, invokes Success Message.
* Clicking "Cancel" or "Close" abandons changes and closes modal without a warning

![image.png](https://images.zenhubusercontent.com/5be0bad045b0290c78fb7458/9007eb6c-2ecc-4e84-8f93-f67d9193d828)

* Secondary Petitioner Contact Info Modal*

* User is informed that Secondary Petitioner contact info must be changed via the Notice of Change of Address form by mail.
* "Notice of Change of Address" links to:
* Clicking "OK" or "Close" closes the modal window

![image.png](https://images.zenhubusercontent.com/5be0bad045b0290c78fb7458/a7b99c78-f924-4177-af2e-8edcc9ffd968)

*Success Message*

* When contact information has been saved, display a success message
* Display updated contact info in the Party Information section for the Primary contact.

![image.png](https://images.zenhubusercontent.com/5be0bad045b0290c78fb7458/c2d84256-fd0c-474b-94ed-2a7decb6f0c3)

**Comment by uxforgood** [comment]
## UX Review Comments/Bugs

- [ ] 1) Font size of all text inside the Case Information "cards" should be 17px - currently appearing as 16px (technically unrelated to this story, so LMK if you'd rather move to a cleanup story)
- [ ] 2) Icon next to "Edit" under Primary contact should be 16.75 wide by 15 height
- [ ] 3) GLOBAL CHANGE - State dropdown on the "Edit" modal (and all state dropdowns for addresses) need to only display the abbreviation for the state, NOT the full name. 
- [ ] 4) On Edit modal - Add Country dropdown to top (per design)
- [ ] 5) On Edit modal - Add Phone Number field to bottom (per design)
- [ ] 6) Secondary Petitioner Contact Info Modal - in the body copy text, "US Tax Court" should be "U.S. Tax Court" 
- [ ] 7) Secondary Petitioner Contact Info Modal - add "Close" function to the modal (Yep, it's repetitive but there's a UX reason to include it) 

- [ ]

**Comment by matthopson** [comment]
@uxforgood Just want to confirm some of the changes before starting on these.

I'm reusing a component for the address form, so the order of the fields is dictated by previous designs where we are capturing contact information. I'll have to make a new component entirely to match this design.

**Comment by uxforgood** [comment]
@matthopson 

The order isn't changing from how it is on the original petition form, etc. My comments are that the Country and Phone Number fields are missing completely from the modal and should be added at the top/bottom as specified in the design.

**Comment by mariahkannenberg** [comment]
### Test Scenarios

**Petitioner Changes Their Personal Information**
Petitioner logs in
Petitioner navigates to case detail>Case information tab
Petitioner selects “Edit” link under primary contact and gets popup
(how to verify pre-population of data?)
Petitioner selects “United States” from Country dropdown
Petitioner enters “100 Main St.” into Mailing Address field
Petitioner enters “Grand View Apartments” into Address Line 2
Petitioner enters “Apt. #104” into Address LIne 3 field
Petitioner enters “Jordan” into City field
Petitioner selects “MN” from State dropdown
Petitioner enters “55352” into ZIP Code field
Petitioner enters “111-111-1111” into Phone Number field
Petitioner selects Save
Petitioner sees success message
Petitioner is directed back to case detail>Case information tab
Petitioner sees under the primary contact label:
	<Petitioner Name>
	100 Main St.
	Grand View Apartments
	Apt. #104
	Jordan, MN 55352
	111-111-1111
	<Petitioner Email Address>
Petitioner logs out

Error scenarios?


**Petitioner Attempts to Change Secondary Contact Information**

Petitioner logs in
Petitioner navigates to case detail>Case information tab
Petitioner selects “Why can’t I edit this?” link and gets popoup
Within the popup, petitioner sees (enter content here?)
Petitioner selects “Notice of CHange of Address” link and PDF of document opens in another window
Petitioner navigates back to portal (Is this even possible to automate?)
Petitioner selects OK button in modal, modal closes and petitioner is directed back to case detail>case information tab
Petitioner logs out

**Comment by sarahridgway** [comment]
Updated specification above to reflect change from a contact information modal to a contact information page. Changes are also reflected in the for devs figma file and were posted to slack and trello.

**Comment by JessicaMarine** [comment]
PO Review 7/12/19:
* When I tried to change the address in 182-19, I got an error message that Address Line 2 (which is optional) cannot be left blank.  This might be because there was previously an address on that line?
![image.png](https://images.zenhubusercontent.com/5bbfb4d092b88e6ddb8b0e3d/5648202d-ef75-412a-b53c-4bd12921fe2b)
